ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java

```diff
@@ -0,0 +1,191 @@
+package com.facebook.react;
+
+import android.app.Activity;
+import android.content.Intent;
+import android.net.Uri;
+import android.os.Build;
+import android.os.Bundle;
+import android.provider.Settings;
+import android.view.KeyEvent;
+import android.widget.Toast;
+
+import com.facebook.common.logging.FLog;
+import com.facebook.infer.annotation.Assertions;
+import com.facebook.react.common.ReactConstants;
+import com.facebook.react.devsupport.DoubleTapReloadRecognizer;
+import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler;
+
+import javax.annotation.Nullable;
+
+/**
+ * A delegate for handling React Application support. This delegate is unaware whether it is used in
+ * an {@link Activity} or a {@link android.app.Fragment}.
+ */
+public class ReactDelegate {
+
+  public final int REQUEST_OVERLAY_PERMISSION_CODE = 1111;
+  private static final String REDBOX_PERMISSION_MESSAGE =
+    "Overlay permissions need to be granted in order for react native apps to run in dev mode.";
+  private static final String REDBOX_PERMISSION_GRANTED_MESSAGE =
+    "Overlay permissions have been granted.";
+
+
+  private final Activity mActivity;
+  private ReactRootView mReactRootView;
+
+
+  @Nullable
+  private final String mMainComponentName;
+
+  @Nullable
+  private Bundle mLaunchOptions;
+
+  @Nullable
+  private DoubleTapReloadRecognizer mDoubleTapReloadRecognizer;
+
+
+  public ReactDelegate(Activity activity, String appKey, Bundle launchOptions) {
+    mActivity = activity;
+    mMainComponentName = appKey;
+    mLaunchOptions = launchOptions;
+    mDoubleTapReloadRecognizer = new DoubleTapReloadRecognizer();
+  }
+
+  public void onHostResume() {
+    if (getReactNativeHost().hasInstance()) {
+      if (mActivity instanceof DefaultHardwareBackBtnHandler) {
```

Comment:
@serj-lotutovici So I updated the code a bit in https://github.com/facebook/react-native/pull/12199/commits/b089982ebe9cec0e86a71bf4986c783ac5355e37

After messing with fragment navigations stacks this week, it seemed easier to allow the Activity to handle the back navigation. If the Fragment handled the back button handler, it ended up getting stuck in a loop, or it wasn't able to safely call the `super.onBackPressed()` like I was hoping for. I'm open to suggestions but this is the conclusion I came to.

This will be shown via documentation for sure, but I'll work on documentation if/when we get this PR "approved" and the idea is solidified.